docs(auth): Mention 403 pitfall + Java specifics (#1463)
cap-js/docs: guides/security/authorization.md

The [restrict annotation](#restrict-annotation) for an entity allows you to enforce authorization checks that statically depend on the event type and user roles. In addition, you can define a `where`-condition that further limits the set of accessible instances. This condition, which acts like a filter, establishes an *instance-based authorization*.

The condition defined in the `where`-clause typically associates domain data with static [user claims](#user-claims). Basically, it *either filters the result set in queries or accepts only write operations on instances that meet the condition*. This means that, the condition applies to following standard CDS events only<sup>1</sup>:
- `READ` (as result filter)
- `UPDATE` (as reject condition<sup>2</sup>)
- `DELETE` (as reject condition<sup>2</sup>)

>  > <sup>1</sup> Node.js supports _static expressions_ that *don't have any reference to the model* such as `where: $user.level = 2` for all events.  

Reviewer: Where is the reference for 1? The list above only refers to 2

The [restrict annotation](#restrict-annotation) for an entity allows you to enforce authorization checks that statically depend on the event type and user roles. In addition, you can define a `where`-condition that further limits the set of accessible instances. This condition, which acts like a filter, establishes an *instance-based authorization*.

The condition defined in the `where`-clause typically associates domain data with static [user claims](#user-claims). Basically, it *either filters the result set in queries or accepts only write operations on instances that meet the condition*. This means that, the condition applies to following standard CDS events only<sup>1</sup>:
- `READ` (as result filter)
- `UPDATE` (as reject condition<sup>2</sup>)
- `DELETE` (as reject condition<sup>2</sup>)

> TODO: But in Java, it also works for custom events? And in Node as well, right?
> 

Author: ```suggestion
```

 > <sup>1</sup> Node.js supports _static expressions_ that *don't have any reference to the model* such as `where: $user.level = 2` for all events.  

>  > <sup>2</sup> CAP Java uses a filter condition by default.

Author: You refer to my TODO for custom events?

Something like:

> - custom event (as reject condition if it selects an entity)

It seems that Node.js assumes that bound actins/functions trigger read events (SELECTs), but I could be mistaken.

>  > <sup>2</sup> CAP Java uses a filter condition by default.

Author: Comment was:

> TODO: But in Java, it also works for custom events? And in Node as well, right?

>  > <sup>2</sup> CAP Java uses a filter condition by default.

Author: IMHO, finished.

* [Exists predicate](#exists-predicate) based on subselects.

<div class="impl java">

> CAP Java 3.5.0 introduced an option to enable rejection conditions for `UPDATE`, `DELETE` and custom events. Enable it via <Config java keyOnly>cds.security.authorization.instance-based.reject-selected-unauthorized-entity.enabled: true</Config>

Reviewer 2: the reject condition also works with READ, doesn't it?

Author: The reject condition is only implemented for write-events (+ custom events). For read events, it remains 404, i.e. filtering.

</div>

> ::: info Avoid enumerable IDs

Reviewer 2: ```suggestion
::: info Avoid enumerable keys
```

> For `UPDATE` and `DELETE`, it is not checked if the user can read the targeted entity. If the user is not authorized, the request is rejected, which could disclose to unauthorized users that an entity under a given key exists.

Reviewer 2: ```suggestion
In case the filter condition is not met in an `UPDATE` or `DELETE` request, the runtime rejects the request (response code 403) even if the user is not even allowed to read the entity. To avoid to disclosure the existence of such entities to unauthorized users, make sure that the key is not efficiently enumerable.
```
